### DISCUSSION TITLE
What's the cleanest way to have every workflow activity log their inputs and outputs upon execution?

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#9126] [Ideas]
Hi,

Currently debugging workflows is hard, and I usually have to log their inputs and outputs to properties to debug the workflow. is there any way to enable general logging for the faults and errors to display in the workflow state json?

I noticed in the workflow module, there is this chunk of code which hints at a future logging feature, where can I learn more about it?

```
    @*TODO: Enable this when workflow logging is implemented.*@
    @*<div id="log" class="tab-pane" role="tabpanel" aria-labelledby="log-tab">
            <div class="card form-group">
                <div class="card-body">
                    Pretty LOG goes here...
                </div>
            </div>
        </div>*@
```

**Comment by weirdyang** [comment]
upon investigation, I realised the workflow executed activities property is never populated. So I’ve implemented my own workflow manager based on the code in the master branch. However I noticed the dev branch has quite a few changes, especially with respect to locking and atomicity. Are those important changes?

At this point I’m saving the snapshot of the state and the execution time as a json string into the workflow executedActivities property and then rendering it for the workflow details page. 

Happy to share my code if anyone is interested.

**Reply by scottiemc7** [reply]
This seems like a bug. I was also looking for the executed activities state property to be updated for debugging purposes, but it's always empty. It's difficult to determine where a workflow faulted or which path it took without this information.
